File: leetcode/0003_longest_substring-without-repeating-characters/longest_substring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = {}
        left, maxVal = 0, 0
        i = 0
        for val in s :
          if val in mp:
                left =  max(left, mp[val] + 1)
          mp[val] = i
          maxVal = max(maxVal, i - left +1)
          i = i + 1

        return maxVal;

solution = Solution()
print(solution.lengthOfLongestSubstring("abcabcbb"))

File: leetcode/0003_longest_substring-without-repeating-characters/longest_substring.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
	fmt.Println(lengthOfLongestSubstring("bbbbb"))
	fmt.Println(lengthOfLongestSubstring("pwwkew"))
}

func lengthOfLongestSubstring(s string) int {
	mp := make(map[byte]int)
	max := 0
	left := 0
	for i := 0; i < len(s); i++ {
		if val, ok := mp[s[i]]; ok {
			left = maxVal(left, val+1)
		}
		mp[s[i]] = i
		max = maxVal(max, i-left+1)
	}
	return max
}

func maxVal(x, y int) int {
	if x < y {
		return y
	}

	return x
}
